src/debug.c

```diff
@@ -2542,41 +2592,48 @@ static int is_thread_ready_to_signal(pid_t pid, pid_t tid, int sig_num) {
  * Writes into tids the tids of these threads.
  * If it fails, returns 0.
 */
-static size_t get_ready_to_signal_threads_tids(pid_t pid, int sig_num, pid_t tids[TIDS_MAX_SIZE]) {
-    /* Initialize the path the process threads' directory. */
-    char path_buff[MAX_BUFF_LENGTH];
-    snprintf(path_buff, MAX_BUFF_LENGTH, "/proc/%d/task", pid);
+static size_t get_ready_to_signal_threads_tids(int sig_num, pid_t tids[TIDS_MAX_SIZE]) {
+    /* Open /proc/<pid>/task file. */
+    char path_buff[PATH_MAX] = "/proc/";
+    _safe_snprintf(path_buff, PATH_MAX, "/proc/%d/task", getpid());
 
-    /* Get the directory handler. */
-    DIR *dir;
-    if (!(dir = opendir(path_buff))) return 0;
+    int dir;
+    if (-1 == (dir = open(path_buff,  O_RDONLY | O_DIRECTORY))) return 0;
 
     size_t tids_count = 0;
-    struct dirent *entry;
     pid_t calling_tid = syscall(SYS_gettid);
     int current_thread_index = -1;
+    long nread;
+    char buff[PATH_MAX];
 
     /* Each thread is represented by a directory */
-    while ((entry = readdir(dir)) != NULL) {
-        if (entry->d_type == DT_DIR) {
+    while ((nread = syscall(SYS_getdents64, dir, buff, PATH_MAX))) {
+        if (nread == -1) {
+            serverLogFromHandler(LL_WARNING, "get_ready_to_signal_threads_tids(): Failed to read the process's task directory");
+            return 0;
+        }
+        for (long pos = 0; pos < nread;) {
+            struct dirent64 *entry = (struct dirent64 *)(buff + pos);
```

Comment:
One thing I'm not sure is if this struct definition is portable. Same link I pointed before: [here](https://github.com/python/cpython/blob/b75186f69edcf54615910a5cd707996144163ef7/Modules/_posixsubprocess.c#L396), it defines `struct linux_dirent64` itself.

Also, there are some bits in **notes** section that implies we are supposed to define ourselves: [link](https://man7.org/linux/man-pages/man2/getdents.2.html)
Probaby, our best bet is defining `struct linux_dirent64` ourselves as in the above link. 

Btw, did you run daily for this branch? I assume our alpine build uses musl libc, it may not provide `struct dirent64`.

Reply:
The fact that we define the struct won't protect us if `syscall(SYS_getdents64)` expects to fill another struct. 
Note that although the man page implies that, the struct mentioned is incorrect. (d_name comes before d_type) 
This is why I feel more comfortable using something defined in a standard header.

and yes
https://github.com/meiravgri/redis/actions/runs/6780556294

Reply:
> Note that although the man page implies that, the struct mentioned is incorrect. (d_name comes before d_type)

Beware, there are two structs: `struct linux_dirent` and `struct linux_dirent64`. Variable order is slightly different in these two (yes, it is confusing!). We are using `struct linux_dirent64` with `syscall(SYS_getdents64)`

The above [link](https://github.com/python/cpython/blob/b75186f69edcf54615910a5cd707996144163ef7/Modules/_posixsubprocess.c#L396) mentiones that this struct definition cannot change without breaking the linux API. So, that's why they defined themselves. Otherwise, we have to assume libc will provide it and libc provided struct will always match the required struct by this syscall. I'm not sure this assumption is better. Considering defining it in the code has worked for these guys, I'd do the same.

Reply:
I know they are not the same (btw SYS_getdents didn't work at all). And that's another thing to consider. I didn't find any documentation that explicitly defines the struct for `SYS_getdents64`. what I did find was glibc docs on 
[low-level directory access](https://www.gnu.org/software/libc/manual/html_node/Low_002dlevel-Directory-Access.html). However `getdents64` function has been supported since glibc 2.2, and some tests failed trying to use it (see https://github.com/redis/redis/pull/12658/commits/6e3fb9959d516100c43b10b6dafdec06794d296c), I believe that an official documentation of glibc is more reliable.

Reply:
This snipped is from [here](https://man7.org/linux/man-pages/man2/getdents.2.html), see the notes section. 
```
HISTORY    
       SVr4.

       getdents64()
              glibc 2.30.
NOTES    
       glibc does not provide a wrapper for getdents(); call getdents()
       using syscall(2)  In that case you will need to define the
       linux_dirent or linux_dirent64 structure yourself.
```

Looks like they added getdents64() in a later version but we can't use it as you showed it already fails on some environments. 

Back to original question, can we rely on existence of `struct dirent64` from libc? I just don't see any guarantees for it. In all example codes ([link](https://linux.die.net/man/2/getdents64)), it defines the struct before caling `syscall()`. Just to be on the safe side, I'd declare it like [here](https://github.com/python/cpython/blob/b75186f69edcf54615910a5cd707996144163ef7/Modules/_posixsubprocess.c#L396).

Reply:
ok